README
pyterm
======

Copyright (c) 2013, Friedrich Paetzke (f.paetzke@gmail.com)

some highlighting functions for terminals.
for usage see src/test.py

## links

* http://ascii-table.com/ansi-escape-sequences-vt-100.php
* http://wiki.ubuntuusers.de/Bash/Prompt


CHANGELOG.org
* CHANGELOG

** 0.1.2
- Switch from README.org to README.rst
  This enables you to install this package directly from a git repository.

** 0.1.1
- Enable usage of foreground and background color at the same time

** 0.1.0
- Add support for PyPy

** 0.0.3
- Enable coloring in builtin /print()/


src/test.py
#! /usr/bin/env python3
'''
@author: Friedrich Paetzke
@copyright: Copyright (c) 2013 Friedrich Paetzke
'''

import pyterm
import time


def main():
    pyterm.fancyPrint('123')
    pyterm.fancyPrint('123 bold', bold=True)
    pyterm.fancyPrint('123 underline', underline=True)
    pyterm.fancyPrint('123 blinking', blinking=True)
    pyterm.fancyPrint('123 bold underline', bold=True, underline=True)
    pyterm.fancyPrint('123 bold blinking', bold=True, blinking=True)
    pyterm.fancyPrint('123 bold underline red', bold=True, underline=True, color=pyterm.Color.Red)
    pyterm.fancyPrint('123 underline green', underline=True, color=pyterm.Color.Green)
    pyterm.fancyPrint('123 brown', color=pyterm.Color.Brown)
    pyterm.fancyPrint('123 brown bold', bold=True, color=pyterm.Color.Brown)

    pyterm.fancyPrint('123 black', color=pyterm.Color.Black)
    pyterm.fancyPrint('123 red', color=pyterm.Color.Red)
    pyterm.fancyPrint('123 green', color=pyterm.Color.Green)
    pyterm.fancyPrint('123 brown', color=pyterm.Color.Brown)
    pyterm.fancyPrint('123 blue', color=pyterm.Color.Blue)
    pyterm.fancyPrint('123 purple', color=pyterm.Color.Purple)
    pyterm.fancyPrint('123 cyan', color=pyterm.Color.Cyan)
    pyterm.fancyPrint('123 ligth grey', color=pyterm.Color.LightGrey)

    pyterm.fancyPrint('123 dark grey', color=pyterm.Color.DarkGrey)
    pyterm.fancyPrint('123 light red', color=pyterm.Color.LightRed)
    pyterm.fancyPrint('123 light green', color=pyterm.Color.LightGreen)
    pyterm.fancyPrint('123 yellow', color=pyterm.Color.Yellow)
    pyterm.fancyPrint('123 light blue', color=pyterm.Color.LightBlue)
    pyterm.fancyPrint('123 light purple', color=pyterm.Color.LightPurple)
    pyterm.fancyPrint('123 light cyan', color=pyterm.Color.LightCyan)
    pyterm.fancyPrint('123 white', color=pyterm.Color.White)


    pyterm.fancyPrint('123 black bg', bgcolor=pyterm.BgColor.Black)
    pyterm.fancyPrint('123 red bg', bgcolor=pyterm.BgColor.Red)
    pyterm.fancyPrint('123 green bg', bgcolor=pyterm.BgColor.Green)
    pyterm.fancyPrint('123 brown bg', bgcolor=pyterm.BgColor.Brown)
    pyterm.fancyPrint('123 blue bg', bgcolor=pyterm.BgColor.Blue)
    pyterm.fancyPrint('123 purple bg', bgcolor=pyterm.BgColor.Purple)
    pyterm.fancyPrint('123 cyan bg', bgcolor=pyterm.BgColor.Cyan)
    pyterm.fancyPrint('123 grey bg', bgcolor=pyterm.BgColor.Grey)

    for i in range(3):
        pyterm.updateLine('123' + str(i))
        time.sleep(0.5)

    pyterm.fancyPrint('')

    for i in range(3):
        s = pyterm.sFancyPrint(str(i) + '123' + str(i), color=pyterm.Color.Red)
        pyterm.updateLine(s)
        time.sleep(0.5)

    pyterm.fancyPrint('')

    for i in reversed(range(101)):
        pyterm.updateLine('123' + str(i))
        time.sleep(0.1)
    pyterm.fancyPrint('')


if __name__ == '__main__':
    main()


test/test.py
#! /usr/bin/env python3
"""
consolor

Copyright (c) 2013, Friedrich Paetzke (f.paetzke@gmail.com)
All rights reserved.

"""
import sys
import os
import unittest

BASE_DIR = os.path.dirname(os.path.abspath(os.path.dirname(__file__)))
sys.path.insert(0, BASE_DIR)

import consolor


class TestConsolor(unittest.TestCase):

    def test_print_bold(self):
        result = consolor.get_line('123 bold', bold=True)
        expected = '\x1b[1m123 bold\x1b[0m'
        self.assertEqual(result, expected)

    def test_print_underline(self):
        result = consolor.get_line('123 underline', underline=True)
        expected = '\x1b[4m123 underline\x1b[0m'
        self.assertEqual(result, expected)

    def test_get_bgcolor(self):
        result = consolor.get_line('123 green bg', bgcolor=consolor.BgColor.Green)
        expected = '\x1b[42;1m123 green bg\x1b[0m'
        self.assertEqual(result, expected)

    def test_get_color(self):
        result = consolor.get_line('123 light green', color=consolor.Color.LightGreen)
        expected = '\x1b[1;32m123 light green\x1b[0m'
        self.assertEqual(result, expected)

    def test_update_line(self):
        for i in reversed(range(101)):
            line = consolor.get_line('123%d' % i, update_line=True)
            expected = '\x1b[2K\r%s%d\x1b[0m' % ('123', i)
            self.assertEqual(line, expected)


if __name__ == '__main__':
    unittest.main()


tests/__init__.py
# -*- coding: utf-8 -*-
"""
consolor

Copyright (c) 2013, Friedrich Paetzke (f.paetzke@gmail.com)
All rights reserved.

"""


src/pyterm.py
'''
@author: Friedrich Paetzke
@copyright: Copyright (c) 2013 Friedrich Paetzke
@license: GPL
'''

_TURN_OFF_CHARACTER_ATTS = '\033[0m'
_TURN_BOLD_MODE_ON = '\033[1m'
_TURN_UNDERLINE_MODE_ON = '\033[4m'
_TURN_BLINKING_MODE_ON = '\033[5m'


class Color:
    Black = '\033[0;30m'
    Red = '\033[0;31m'
    Green = '\033[0;32m'
    Brown = '\033[0;33m'
    Blue = '\033[0;34m'
    Purple = '\033[0;35m'
    Cyan = '\033[0;36m'
    LightGrey = '\033[0;37m'
    DarkGrey = '\033[1;30m'
    LightRed = '\033[1;31m'
    LightGreen = '\033[1;32m'
    Yellow = '\033[1;33m'
    LightBlue = '\033[1;34m'
    LightPurple = '\033[1;35m'
    LightCyan = '\033[1;36m'
    White = '\033[1;37m'


class BgColor:
    Black = '\033[40;1m'
    Red = '\033[41;1m'
    Green = '\033[42;1m'
    Brown = '\033[43;1m'
    Blue = '\033[44;1m'
    Purple = '\033[45;1m'
    Cyan = '\033[46;1m'
    Grey = '\033[47;1m'


def fancyPrint(s, bold=False, underline=False, blinking=False, color=None,
               bgcolor=None, end='\n'):
    '''
    prints `s` with the given formatting

    @type s: str
    @type bold: bool
    @type underline: bool
    @type blinking: bool
    @type color: str
    @type bgcolor: str
    @type end: str
    '''
    s = sFancyPrint(s, bold=bold, underline=underline, blinking=blinking,
                    color=color, bgcolor=bgcolor)
    print(s, end=end)


def sFancyPrint(s, bold=False, underline=False, blinking=False, color=None,
                bgcolor=None):
    '''
    returns `s` with the given formatting

    @type s: str
    @type bold: bool
    @type underline: bool
    @type blinking: bool
    @type color: str
    @type bgcolor: str
    @return: str
    '''
    fmt = ''
    for val in [bgcolor, color]:
        if val:
            fmt += val

    if bold:
        fmt += _TURN_BOLD_MODE_ON
    if underline:
        fmt += _TURN_UNDERLINE_MODE_ON
    if blinking:
        fmt += _TURN_BLINKING_MODE_ON

    return fmt + s + _TURN_OFF_CHARACTER_ATTS


def updateLine(s):
    '''
    deletes the output of the current line and prints `s` on the same line
    without a new line

    @type s: str
    '''
    print('\033[2K\r' + s, end='')


consolor/consolor.py
# -*- coding: utf-8 -*-
"""
consolor

Copyright (c) 2013-2014, Friedrich Paetzke (f.paetzke@gmail.com)
All rights reserved.

"""
from __future__ import print_function


_TURN_OFF_CHARACTER_ATTS = '\033[0m'
_TURN_BOLD_MODE_ON = '\033[1m'
_TURN_UNDERLINE_MODE_ON = '\033[4m'
_TURN_BLINKING_MODE_ON = '\033[5m'
_UPDATE_LINE = '\033[2K\r'


class Color:
    Black = '\033[0;30m'
    Red = '\033[0;31m'
    Green = '\033[0;32m'
    Brown = '\033[0;33m'
    Blue = '\033[0;34m'
    Purple = '\033[0;35m'
    Cyan = '\033[0;36m'
    LightGrey = '\033[0;37m'
    DarkGrey = '\033[1;30m'
    LightRed = '\033[1;31m'
    LightGreen = '\033[1;32m'
    Yellow = '\033[1;33m'
    LightBlue = '\033[1;34m'
    LightPurple = '\033[1;35m'
    LightCyan = '\033[1;36m'
    White = '\033[1;37m'
    Reset = _TURN_OFF_CHARACTER_ATTS


class BgColor:
    Black = '\033[40;1m'
    Red = '\033[41;1m'
    Green = '\033[42;1m'
    Brown = '\033[43;1m'
    Blue = '\033[44;1m'
    Purple = '\033[45;1m'
    Cyan = '\033[46;1m'
    Grey = '\033[47;1m'
    Reset = _TURN_OFF_CHARACTER_ATTS


def print_line(s, bold=False, underline=False, blinking=False, color=None, bgcolor=None, end='\n'):
    """
    Prints a string with the given formatting.

    """
    s = get_line(s, bold=bold, underline=underline,
                 blinking=blinking, color=color, bgcolor=bgcolor)
    print(s, end=end)


def get_line(s, bold=False, underline=False, blinking=False, color=None,
             bgcolor=None, update_line=False):
    """
    Returns a string with the given formatting.

    """
    parts = []

    if update_line:
        parts.append(_UPDATE_LINE)

    for val in [bgcolor, color]:
        if val:
            parts.append(val)

    if bold:
        parts.append(_TURN_BOLD_MODE_ON)
    if underline:
        parts.append(_TURN_UNDERLINE_MODE_ON)
    if blinking:
        parts.append(_TURN_BLINKING_MODE_ON)

    parts.append(s)
    parts.append(_TURN_OFF_CHARACTER_ATTS)

    result = ''.join(parts)
    return result


def update_line(s, bold=False, underline=False, blinking=False, color=None, bgcolor=None):
    """
    Overwrites the output of the current line and prints s on the same line
    without a new line.

    """
    s = get_line(s, bold=bold, underline=underline, blinking=blinking,
                 color=color, bgcolor=bgcolor, update_line=True)
    print(s, end='')
